bindings/rust/src/lib.rs

```diff
@@ -38,95 +30,72 @@ impl fmt::Display for ZenError {
     }
 }
 
-impl std::error::Error for ZenError {}
+pub fn zencode_exec_extra(
+    script: &str,
+    conf: &str,
+    keys: &str,
+    data: &str,
+    extra: &str,
+    context: &str,
+) -> Result<ZenResult, ZenError> {
+    let mut zen_input: String = "".to_owned();
 
-type ZenGIL = MutexGuard<'static, ()>;
+    zen_input.push_str(conf);
+    zen_input.push_str("\n");
 
-fn aquire_zen_gil() -> ZenGIL {
-    static mut MUTEX: *const Mutex<()> = std::ptr::null();
-    static ONCE: Once = Once::new();
-    let mutex: &Mutex<()> = unsafe {
-        ONCE.call_once(|| {
-            MUTEX = std::mem::transmute(Box::new(Mutex::new(())));
-        });
-        MUTEX.as_ref().unwrap()
-    };
-    mutex.lock().unwrap()
-}
+    zen_input.push_str(&general_purpose::STANDARD.encode(script));
+    zen_input.push_str("\n");
 
-const BUF_SIZE: usize = 2 * 1024 * 1024;
+    zen_input.push_str(&general_purpose::STANDARD.encode(&keys));
+    zen_input.push_str("\n");
 
-type Fun = unsafe extern "C" fn(
-    *const ::std::os::raw::c_char,
-    *const ::std::os::raw::c_char,
-    *const ::std::os::raw::c_char,
-    *const ::std::os::raw::c_char,
-    *mut ::std::os::raw::c_char,
-    ::std::os::raw::c_ulong,
-    *mut ::std::os::raw::c_char,
-    ::std::os::raw::c_ulong,
-) -> ::std::os::raw::c_int;
+    zen_input.push_str(&general_purpose::STANDARD.encode(&data));
+    zen_input.push_str("\n");
 
-pub fn zencode_exec(
-    script: impl AsRef<str>,
-    conf: impl AsRef<str>,
-    keys: impl AsRef<str>,
-    data: impl AsRef<str>,
-) -> Result<ZenResult, ZenError> {
-    exec_f(c::zencode_exec_tobuf, script, conf, keys, data)
+    zen_input.push_str(&general_purpose::STANDARD.encode(&extra));
+    zen_input.push_str("\n");
+
+    zen_input.push_str(&general_purpose::STANDARD.encode(&context));
```

Comment:
shouldn't we define base64 as the encoding here, rather than relying on a standard which may as well change?